docs/rfcs/004-session-management.md

```diff
@@ -0,0 +1,125 @@
+
+> There are things that are serialized and stored when Atom exits. It is completely non-obvious what items belong to this set and what items do not as well as how to restore the saved items when next launching Atom. Decisions should be made about whether some of these items belong to the specific mode that Atom is in or if they are more global state that people expect to be in place the next time Atom is launched (or even an individual file is opened post-launch) no matter what mode Atom is launched in.
+
+Let's revisit the way we manage session state so that it becomes an Atom feature that users can understand and trust.
+
+## Motivation
+
+_pending_
+
+Problems to address:
+
+* The active project (which is deserialized on open, serialized on close) is chosen implicitly based on the set of project folders.
+* The active project isn't explicitly named anywhere in the UI. Two open Atom windows with two different projects do not obviously differ; two open Atom windows on the same project are not obviously the same.
+* Adding or removing a root folder silently changes the active project.
+* When the same project is open in two Atom windows at once, the last-closed window clobbers any project state saved by the first-closed.
+* Unsaved content in a TextBuffer is only accessible through the project that was open when it was created.
+
+## Explanation
+
+A _Session_ is the unit of persistent window state in Atom related to a coherent development effort. It includes:
+
+* The set of root folders;
+* Pane and dock configuration;
+* Open pane and dock items;
+* Cursor position, fold state, and other TextEditor characteristics;
+* Per-package [serialized state](https://flight-manual.atom.io/behind-atom/sections/serialization-in-atom/).
+
+Each Atom window is associated with one Session, and each Session is associated with at most one Atom window at a time.
+
+Each Session has a unique, user-customizable name.
+
+Other persistent state is tracked _per-file_, based on its inode or fileID to work through filesystem softlinks or hardlinks. This includes:
+
+* Unsaved buffer changes;
+* Per-file manually changed indentation, line ending, and encoding settings.
+
+### Add an explicit indication of the current Session to the tree-view
+
+At the top of the tree-view, display the current Project's name. Clicking it opens the project management UI described below.
+
+> TODO: graphic
+
+### Add a command-line argument to explicitly choose the Session
+
+When opening Atom from the command line with one or more locations specified:
+
+1. If exactly one non-open Session matches those paths, that Session is loaded into the newly opened Atom window.
+2. If no non-open Sessions match, a new Session is created for that Atom window.
+3. If multiple Sessions match, a new Session is created for that Atom window, and a notification is displayed allowing the user to switch to any of the matching Sessions and destroy the just-created one.
```

Comment:
Why allow multiple sessions of the same name? I'd expect the names to be unique.

Reply:
Session names are unique, but multiple sessions may contain the same project folders. (For an example of why this may be useful, I often have two Atom windows open on the project I'm working, one for "reference" and one for "input".)

For example, given:

```sh
atom --session one project-a/ project-b/
atom --session two project-b/
atom --session three project-b/ project-c/
```

If you run:

```sh
atom project-b/
```

Atom would:

1. Open a new window containing `project-b/`
2. Create a new session (say "session-123")
3. Display a notification offering to switch to sessions "one", "two", or "three".
   * If the user chooses an existing session -- say "three" -- the current window loads the chosen session (so project folders `project-b/` and `project-c/`, any open editors, pane arrangement and so on) and "session-123" is destroyed
  * If the user dismisses the notification, "session-123" remains in use.

Does that make sense?

Reply:
Yea that's more clear to me. 👍